### ISSUE TITLE
Make class Fl_Simple_Terminal  refuse the focus

### ISSUE THREAD

**Issue by ManoloFLTK** [fltk/fltk#365]
I believe class Fl_Simple_Terminal should implement the handle(int) virtual function
to refuse the focus when it receives the FL_FOCUS event :
```
int Fl_Simple_Terminal::handle(int event) {
  if (event == FL_FOCUS) return 0;
  return Fl_Text_Display::handle(event);
}
```
The problem is visible for example with test/input. Add :
`printf("Fl::focus()=%p Fl_Simple_Terminal=%p\n", Fl::focus(), G_tty);`
after the `window->show(argc,argv); `statement at the end
shows that the focus is at the terminal when input is launched, which makes no sense
since the doc of Fl_Simple_Terminal states :
> What this widget is NOT is a full terminal emulator; it does NOT handle stdio redirection, pipes, pseudo ttys, termio character cooking, **keyboard input processing**,…

@erco77 Do you agree?

**Comment by Albrecht-S** [comment]
I believe a better solution would be to use `clear_visible_focus()` in the constructor of `Fl_Simple_Terminal` which is documented to exclude the widget from keyboard navigation and achieves the same effect WRT setting the initial focus.

Unfortunately this (or `G_tty->visible_focus(0);` in `test/input.cxx`) does not work because this disables setting the focus to the `Fl_Simple_Terminal` widget completely which makes it impossible to copy text from the terminal to the clipboard. That's bad. However...

Looking at the code of `Fl::focus(Fl_Widget *)` reveals that there is IMHO a bug because it refuses to set the focus to a widget by `if (o && !o->visible_focus()) return;`. Commenting out this statement and using `clear_visible_focus()` in the constructor of `Fl_Simple_Terminal` works as expected.

I filed another issue (#366) for further investigation.

**Comment by Albrecht-S** [comment]
@ManoloFLTK The effect of setting the wrong initial focus in this particular test program (`test/input.cxx`) is caused by adding the `Fl_Simple_Terminal` widget as the first widget in the group (window) which is weird by itself. This should be fixed anyway by moving the terminal widget to the end.

Was this demo program the (only) reason why you filed this issue or did you see similar effects in other programs as well?

**Comment by ManoloFLTK** [comment]
So, there's another aspect to the problem: without focus, selection and copy of text may become impossible.
I'll look into that and into issue #366.

@Albrecht-S I've been using test/input (and test/editor) to develop support for text input methods under Wayland, and was annoyed the focus was not straight in one of the text fields of that program.

**Comment by erco77** [comment]
Yes, the only thing I can think of why it would need it is for keyboard navigation of text selection, but that's turned off -- I think its presence was unexpected + distracting in a terminal context.

I did a quick test with the handle() override, and it doesn't seem to affect navigation of the scrollbars with PgUp/Dn and Up/Down arrows (which are I think handled by non-focus oriented FL_SHORTCUT events)

So should be good.

I didn't look into Albrecht's method (pending possible bug fix), but that might work too, assuming it too doesn't affect the above.

**Comment by erco77** [comment]
> So, there's another aspect to the problem: without focus, selection and copy of text may become impossible.

Hmm, in my test with the overridden handle() and the test/input demo, after using the 'print_changed()' button to cause text in the simple terminal, I can highlight text, hit ^C, and then use right-click Paste into a terminal. At least on linux.

But I'm not sure by what means that is working though.. perhaps when you make a text selection, it becomes the focus widget anyway, even though it rejects FL_FOCUS events? I didn't dig, just guessing..

**Comment by ManoloFLTK** [comment]
> Hmm, in my test with the overridden handle() and the test/input demo, after using the 'print_changed()' button to cause text in the simple terminal, I can highlight text, hit ^C, and then use right-click Paste into a terminal. At least on linux.

That's also what I see now with macOS.
So, focus and text selection seem decoupled, which is perfect for the simple terminal wdget.

**Comment by Albrecht-S** [comment]
@ManoloFLTK @erco77 Yes, I think that "focus and text selection seem decoupled" is correct and on Linux text selection alone copies the text to the selection clipboard and can be used to paste it -- in my setup by clicking the middle mouse button.

But that's unique to Unix / X11. On other platforms you need to type `^C` as Greg described and this needs the focus to deliver the keypress to the widget.

Side note: it **might** be possible that a widget still gets the keypress as an `FL_SHORTCUT` event but this would be unreliable at least (if at all possible). I wouldn't want to pursue this (i.e. check the sources of `Fl_Text_Display`) but we should be aware that it **might** be possible.

Hint: A good test tool is `test/clipboard.cxx` which works on all three major platforms and displays the copied text instantaneously. I used this for my tests.

**Comment by erco77** [comment]
> But that's unique to Unix / X11. On other platforms you need to type ^C as Greg described and this needs the focus to deliver the keypress to the widget.

Just to clarify, I was wondering aloud if ^C worked _perhaps_ because during its handling of the highlighting process, if it were to set itself to be the focus widget with Fl::focus(), despite ignoring FL_FOCUS events, becomes the focus widget anyway, and thus the ^C event is thrown to the Fl_Text_Xxx handler as a regular FL_KEYBOARD event by the fact it is still the Fl::focus() widget. Just a guess, but I think it can work that way.

**Comment by Albrecht-S** [comment]
OK, to clarify, we have two possible situations after one of the two suggested patches:

1. Manolo's patch rejects the `FL_FOCUS` event and disables setting the focus to the widget by `TAB` navigation and by the initial focus widget selection. If you click into the terminal widget `Fl_Text_Display` handles the **`FL_PUSH`** event and calls `Fl::focus(this)` which sets the focus despite the fact that it _would be rejected by normal TAB navigation_ (`FL_FOCUS` returns 0). This is what `Fl_Text_Display` always does and it works here too.

2. My suggestion to use `visible_focus(0)` or the equivalent `clear_visible_focus()` **should** IMHO work as expected but does not (I tested it). This is obviously due to the fact that `Fl::focus(Fl_Widget*)` does **NOT** set the focus widget if `VISIBLE_FOCUS` has been disabled, and this is the reason why I opened issue #366.

Summary: I would prefer using `clear_visible_focus()` which doesn't work currently. If we could fix this (see #366) then I'd prefer my proposal. Otherwise we could still use Manolo's suggestion which is generally not wrong (IMO) but it would introduce a handle() method in the derived class which we could avoid.
